### DISCUSSION TITLE
Persisting user-created objects with ElasticSearch

### DISCUSSION THREAD

**Discussion by algadoc** [nasa/openmct#4812] [Community Q&A]
Hi, I am using OpenMCT to view telemetry from a sounding rocket and was wondering how I can persist user-created views in something other than a JSON or local storage. 

There is an ElasticSearch plugin to persist these objects, but I am at a loss on how to implement it. Do I need to create a new object and composition providers that use this plugin to store user-created views in the ElasticSearch server, and if so how can I access these objects through the openMCT API? And, if this is not the right way to do it, how can I go about implementing this plugin?

Thank you for your help!

**Comment by andyjsammons** [comment]
@algadoc Sorry you haven't gotten an answer yet. Did you find a solution? I'm working on a similar project. I want to integrate openMCT with elasticsearch for some basic test cases via an arduino and some sensors. Let me know if you have any insight! I'm ultra-noob level, so I'm open to any advice what so ever, as I can't argue that I understand your question all together. Cheers!

**Comment by akhenry** [comment]
@shefalijoshi

**Comment by shefalijoshi** [comment]
Hi @algadoc !
Open MCT comes with a CouchDB plugin included and instructions to get started using docker can be found here: https://github.com/nasa/openmct/blob/master/src/plugins/persistence/couch/README.md

Please let me know if this works for your use case. 

As you mentioned, alternative could also be to write your own persistence store object provider. 
The required methods you'll need to implement are `get(identifier)`, `create(model)` and `update(model)`. 
Optionally you could also implement the `search()` and `observe(identifier, callback)` methods to extend search and object synchronization functionality.
Our included CouchDB object provider can serve as a guide: https://github.com/nasa/openmct/blob/master/src/plugins/persistence/couch/CouchObjectProvider.js

**Comment by akhenry** [comment]
@algadoc There was at one time an Open MCT plugin to support ElasticSearch for object persistance. It did not support telemetry from ES. This was deprecated because we stopped using it internally. As @shefalijoshi  mentioned a new persistence plugin could be created (and would be very welcome!) to support Elasticsearch, with [the legacy code used as a guide](https://github.com/nasa/openmct/tree/505796d9f0afb095aacd92cdcbf70cd4e5ef6c30/platform/persistence/elastic/src).
